did_resolver_sov/src/resolution/utils.rs

```diff
@@ -0,0 +1,181 @@
+    did_parser::ParsedDID,
+    shared_types::did_document_metadata::DIDDocumentMetadata,
+    traits::resolvable::{
+        resolution_metadata::DIDResolutionMetadata, resolution_output::DIDResolutionOutput,
+    },
+};
+use serde_json::Value;
+
+use crate::{
+    error::{parsing::ParsingErrorSource, DIDSovError},
+    service::{DidSovServiceType, EndpointDidSov},
+};
+
+fn prepare_ids(did: &str) -> Result<(Uri, ParsedDID), DIDSovError> {
+    let service_id = Uri::new(did)?;
+    let ddo_id = ParsedDID::parse(did.to_string())?;
+    Ok((service_id, ddo_id))
+}
+
+fn get_data_from_response(resp: &str) -> Result<Value, DIDSovError> {
+    let resp: serde_json::Value = serde_json::from_str(resp)?;
+    match &resp["result"]["data"] {
+        Value::String(ref data) => serde_json::from_str(data).map_err(|err| err.into()),
+        Value::Null => Err(DIDSovError::NotFound("DID not found".to_string())),
+        resp => Err(DIDSovError::ParsingError(
+            ParsingErrorSource::LedgerResponseParsingError(format!(
+                "Unexpected data format in ledger response: {resp}"
+            )),
+        )),
+    }
+}
+
+fn get_txn_time_from_response(resp: &str) -> Result<i64, DIDSovError> {
+    let resp: serde_json::Value = serde_json::from_str(resp)?;
+    let txn_time = resp["result"]["txnTime"]
+        .as_i64()
+        .ok_or(DIDSovError::ParsingError(
+            ParsingErrorSource::LedgerResponseParsingError("Failed to parse txnTime".to_string()),
+        ))?;
+    Ok(txn_time)
+}
+
+fn posix_to_datetime(posix_timestamp: i64) -> Option<DateTime<Utc>> {
+    NaiveDateTime::from_timestamp_opt(posix_timestamp, 0)
+        .map(|date_time| DateTime::<Utc>::from_utc(date_time, Utc))
+}
+
+pub(super) fn is_valid_sovrin_did_id(id: &str) -> bool {
+    if id.len() < 21 || id.len() > 22 {
+        return false;
+    }
+    for c in id.chars() {
+        match c {
+            '1'..='9' | 'A'..='H' | 'J'..='N' | 'P'..='Z' | 'a'..='k' | 'm'..='z' => {}
+            _ => return false,
+        }
+    }
+    true
+}
+
+pub(super) async fn resolve_ddo(did: &str, resp: &str) -> Result<DIDResolutionOutput, DIDSovError> {
+    let (service_id, ddo_id) = prepare_ids(did)?;
+
+    let service_data = get_data_from_response(resp)?;
+    let endpoint: EndpointDidSov = serde_json::from_value(service_data["endpoint"].clone())?;
+
+    let txn_time = get_txn_time_from_response(resp)?;
+    let datetime = posix_to_datetime(txn_time);
+
+    let service = {
+        let mut service_builder =
+            Service::builder(service_id, endpoint.endpoint.as_str().try_into()?)?;
+        for t in endpoint.types {
+            if t != DidSovServiceType::Unknown {
+                service_builder = service_builder.add_service_type(t.to_string())?;
+            };
+        }
+        service_builder.build()?
+    };
+
+    let ddo = DIDDocument::builder(ddo_id).add_service(service).build();
```

Comment:
Unless I am missing something, the DDO produced by this function is incomplete. Verification method is not set, and `resp` argument passed in argument is ledger response for getting attribute, but you didn't fetch DID's verkey anywhere

Reply:
It also seems that if ledger contains routing_keys or recipient keys, it wouldn't get set up in service object's extras